rust/private/rustc.bzl

```diff
@@ -864,11 +871,31 @@ def _error_format_impl(ctx):
     return [ErrorFormatInfo(error_format = raw)]
 
 error_format = rule(
-    doc = (
-        "A helper rule for controlling the rustc " +
-        "[--error-format](https://doc.rust-lang.org/rustc/command-line-arguments.html#option-error-format) " +
-        "flag."
-    ),
+    doc = ("""
+A helper rule for controlling the rustc
+[--error-format](https://doc.rust-lang.org/rustc/command-line-arguments.html#option-error-format)
+flag. Valid values match the documented values of --error-format.
+
+Typical usage is to pass `--@rules_rust//:error_format=<format>` to `bazel build`, e.g. for
+machine-readable output, `--@rules_rust//:error_format=json`.
+"""),
     implementation = _error_format_impl,
     build_setting = config.string(flag = True),
 )
+
+def _extra_codegen_impl(ctx):
+    return ExtraCodegenInfo(extra_codegen = ctx.build_setting_value)
+
+extra_codegen = rule(
+    doc = ("""
+A helper rule for passing [--codegen](https://doc.rust-lang.org/rustc/codegen-options/index.html)
+parameters to rustc for ALL targets, such as `lto=thin`. Codegen parameters that apply only to
```

Comment:
Specifically for `lto=thin`, I think we could also support the rules_cc way of controlling that - using `--features=thin_lto`. Features are accessible through https://docs.bazel.build/versions/master/skylark/lib/ctx.html#features, and we could query the C++ toolchain to see if it supports thin_lto. In a separate PR is fine, and I'm happy to chat more about this.

Reply:
I looked at that, but I thought that --features=lto was specifically a rules_cc thing. Also if we use the same flag as rules_cc, it becomes impossible to control them individually, and there are more options than just "thin" for Rust.

My actual goal was to get `--codegen=linker-plugin-lto` working, but I haven't gotten there yet because it requires linker plugins to work correctly, and `--codegen=lto=thin` is a simpler thing to get working with similar benefit, so merely supporting --features=thin_lto wouldn't suffice.

Reply:
Internally `--features=thin_lto` affects all languages in the build. Do you anticipate the need to say build rust with thinlto but not C++, in a single build?

There are more options than just "thin" in Rust, but only "thin" is compatible with rules_cc, and "thin" is so much faster to build with Bazel thatI don't think full LTO is going to be used often.

All that said, I'm fine with having rust-specific lto switch, I was proposing to in addition also look into rules_cc's thing and enable thin lto when they do.

I'm not sure I follow the last paragraph, why doesn't `--features=thin_lto` suffice to enable thin lto?

Reply:
I was under the impression from https://doc.rust-lang.org/rustc/linker-plugin-lto.html that `--codegen=lto=thin` is not the correct way to do LTO over C++ & Rust together. That page seems to imply that you need to use the `--codegen=linker-plugin-lto` in Rust and the `-flto=thin` flag in C++.

The docs aren't super clear about whether this is mutually exclusive with `--codegen=lto=thin` though, so maybe we can still respond to `--features=thin_lto` by setting the `--codegen=linker-plugin-lto` _and_ `--codegen=lto=thin`, but I haven't tried it. I just found a vague mention here that they might be compatible https://doc.rust-lang.org/rustc/codegen-options/index.html#linker-plugin-lto